logic/scorer/scorer.go

```diff
@@ -60,260 +34,247 @@ func points(lead model.Card, hand []model.Card, isCrib bool) int {
-	if len(ptVals) == 0 {
-		return 0
+func (ps *pairScorer) atEach(c model.Card) {
+	ps.valuesToCounts[c.Value]++
+	switch ps.valuesToCounts[c.Value] {
+	case 2:
+		ps.points += 2
+	case 3:
+		// we've already added 2 for this value
+		ps.points += 4
+	case 4:
+		// we've already added 6 for this value
+		ps.points += 6
 	}
+}
 
-	var many uint
-	for i, o := range ptVals {
-		if o > goal {
-			break
-		} else if o == goal {
-			many++
-		} else {
-			// o is less than the goal. See what we can find with it
-			subWith := howManyAddUpTo(goal-o, ptVals[i+1:])
-			many += subWith
-		}
+func (ps *pairScorer) accumulate() (scoreType, uint8) {
+	var st scoreType
+	switch ps.points {
+	case 2:
+		st = onepair
+	case 4:
+		st = twopair
+	case 6:
+		st = triplet
+	case 8:
+		st = triplet | onepair
+	case 12:
+		st = quad
 	}
-
-	return many
+	return st, ps.points
 }
 
-// Assumes input is sorted and has len 5
-func scoreRunsAndPairs(values []int) (scoreType, int) { //nolint:gocyclo
-	min := values[0]
-	max := values[4]
+type flushScorer struct {
+	isHandFlush bool
+	isCrib      bool
+	firstCard   model.Card
+	leadCard    model.Card
+}
 
-	// check quad for all hands is the same
-	if values[3] == min || values[1] == max {
-		// this is a quad of the either form
-		// Y Y Y Y|X (wlog cuz Y|X X X X)
-		return quad, 12 /* 4 of a kind is worth 12 */
+func newFlushScorer(lead model.Card, hand []model.Card, isCrib bool) flushScorer {
+	return flushScorer{
+		isHandFlush: true, // innocent until proven guilty
+		isCrib:      isCrib,
+		firstCard:   hand[0],
+		leadCard:    lead,
 	}
+}
 
-	diffOf5Cards := max - min
-
-	if diffOf5Cards == 1 {
-		// this hand is either
-		// A A A|B B (wlog cuz A A|B B B)
-		// A A A A|B (wlog cuz A|B B B B)
-		// but we've already checked the quad
-		// therefore this is a triple and a pair
-		return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-	} else if diffOf5Cards == 2 {
-		// check hands with triples
-		if values[1] == values[3] {
-			// we know that values[0] = values[1] - 1 and values[3] = values[4] + 1
-			// because we already checked quads
-			// A|B B B|C
-			return tripleRunOfThree, 15 /* triple run of 3 */
-		} else if values[2] == min {
-			// A A A|B|C
-			// A A A|C C
-			if values[3] == max {
-				// triple/pair combo
-				return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-			}
-			// triple run of 3 is worth 15
-			// because 3 runs of three (worth 9) and a triple for 6
-			return tripleRunOfThree, 15 /* triple run of 3 */
-		} else if values[2] == max {
-			// A|B|C C C
-			// A A|C C C
-			if values[1] == min {
-				return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-			}
-			return tripleRunOfThree, 15 /* triple run of 3 */
-		}
-
-		// If the hand doesn't have a quad or a triple, then it must be a double-double run of three
-		return doubleDoubleRunOfThree, 16 /* double double run of 3 */
+func (fs *flushScorer) atEach(c model.Card) {
```

Comment:
:cactus: mismatch of types: either this should return a pointer from the constructor or the methods should have struct receivers.

Reply:
My goal was not to escape to the heap and _also_ not copy a whole bunch of data 6 times. However, we actually don't escape to the heap returning a pointer from the constructor (somehow).

As a point of discussion, though, why is it a negative to let go automagically determine whether to pointerize something or not? Would you argue that a type should have _either_ value receivers _or_ pointer receivers, but not both?

Reply:
it doesn't escape because go hasn't assigned a high enough complexity in compilation.

for discussion: yes. A struct should have _all_ pointer receivers and pointer returns from constructors **OR** _all_ value receivers and value returns from constructors. I'm very much in the bill kennedy camp on this issue.

The only exception is for things like json marshal/unmarshal which may need to break that rule.